docs/design-archive/runtime-manager-design-doc.md

```diff
@@ -0,0 +1,116 @@
+priorities under hybrid scenario, applying new isolation policies for latest Linux kernel, CPU architecture, and etc.
+
+There are two components involved, RuntimeManager and RuntimePlugins.
+
+![image](../images/runtime-manager-architecture.svg)
+
+## Goals
+
+- Enhance resource management for scenario like hybridization.
+- Provide interface for new isolation features which are not supported by CRI.
+
+## Components
+
+### RuntimeManager
+
+RuntimeManager is in charge of intercepting request during pod's lifecycle, such as RunPodSandbox, CreateContainer etc.,
+and then calling RuntimePlugins to do resource isolation policies before transferring request to backend containerd(dockerd)
+and after transferring response to kubelet. RuntimeManager provides an isolation-policy-execution framework which allows
+customized plugins registered to do specified isolation policies, these plugins are called RuntimePlugins.
+RuntimeManager itself does NOT do any isolation policies.
+
+### RuntimePlugins
+
+RuntimePlugins register events(RunPodSandbox etc.) to RuntimeManager and would receive notifications when events happen.
+RuntimePlugins should complete resource isolation policies basing on the notification message, and then response
+RuntimeManager, RuntimeManager would decide to transfer request to backend containerd or discard request according to
+plugins' response.
+
+If no RuntimePlugins registered, RuntimeManager would become a transparent proxy between kubelet and containerd.
+
+## Architecture
+
+![image](../images/runtime-manager-design.svg)
+
+There are 4 main components for RuntimeManager.
+
+### CRI Server
+
+As a proxy between kubelet and containerd, RuntimeManager acts as a CRI server for kubelet(http server under dockershim
+scenario). It should intercept all request from kubelet, and generate protocols for talking with plugins before and
+after talking with backend containerd(dockerd)
+
+### Plugins Manager
+
+PluginsManager is in charge of parsing registered plugin info from `/etc/runtime/hookserver.d` dynamically.
+
+### Runtime Dispatcher
+
+RuntimeDispatcher is designed to manage communications with plugins.
+
+### Store
+
+As a proxy, RuntimeManager had better be designed as stateless, but sometimes it does NOT work. Take StartContainer hook
+for example, there exists only containerID in CRI StartContainerRequest, which is not enough for plugins to adapt policies
+since plugins may not store pod/container info(such as meta, priority) locally. So RuntimeManager should store pod/container
+info during RunPodSandbox/CreateContainer Stage. When StartContainer request comes, RuntimeManager can get pod/container info
+by containerID, and then call plugins with pod/container info.
+
+With store, there would be pod/container info everytime RuntimeManager calls plugins, so there is no need for plugins to
+store pod/container info exceptionally, plugins can be designed as stateless.
```

Comment:
Add a consideration about performance, store is in memory and does not generate external io to disk.

Reply:
> Add a consideration about performance, store is in memory and does not generate external io to disk.

changed to: 
Considering performance, store locates in memory and does not generate external io to disk.